zigpy/application.py

```diff
@@ -863,6 +871,59 @@ def get_device_with_address(
         else:
             raise ValueError(f"Invalid address: {address!r}")
 
+    @contextlib.contextmanager
+    def callback_for_response(
+        self,
+        src: zigpy.device.Device,
+        filters: list[zigpy.typing.MatcherType],
+        callback: typing.Callable[
+            [
+                zigpy.zcl.foundation.ZCLHeader | zdo_types.ZDOHeader,
+                zigpy.zcl.foundation.CommandSchema,
+            ],
+            typing.Any,
+        ],
+    ) -> typing.Any:
+        """
+        Context manager to create a callback that is passed Zigbee responses.
+        """
+
+        listener = zigpy.util.CallbackListener(
+            device=src,
+            matchers=tuple(filters),
+            callback=callback,
+        )
+
+        self._req_listeners[src].append(listener)
+
+        try:
+            yield
+        finally:
+            self._req_listeners[src].remove(listener)
+
+    @contextlib.contextmanager
+    def wait_for_response(
+        self,
+        src: zigpy.device.Device,
+        filters: list[zigpy.typing.MatcherType],
+    ) -> typing.Any:
+        """
+        Context manager to wait for a Zigbee response.
+        """
+
+        listener = zigpy.util.FutureListener(
+            device=src,
+            matchers=tuple(filters),
+            future=asyncio.get_running_loop().create_future(),
```

Comment:
why not just `asyncio.Future()` as opposed to `asyncio.get_running_loop().create_future()`

Reply:
It's recommended by [the docs](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future):

> The rule of thumb is to never expose Future objects in user-facing APIs, and the recommended way to create a Future object is to call [loop.create_future()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_future). This way alternative event loop implementations can inject their own optimized implementations of a Future object.

Probably makes no difference.

Reply:
TIL… cool thx :)